src/traits.rs

```diff
@@ -107,6 +106,12 @@ where
         self.messages.extend(other.messages);
     }
 
+    /// Extends this step with `other` and returns the result.
+    pub fn and(mut self, other: Self) -> Self {
```

Comment:
Why do we need this in addition to `extend_with`?

Reply:
Isn't `and` very misleading here? There's no "falsy" `Step`, is there?

Reply:
I have trouble coming up with a better name. `union`? `with`?
I'm also happy with not adding it at all, if you think it's not worth it. I just found so many places where we would first `extend` and then return that I thought it might be handy.

Reply:
Probably not `union` because `extend` is not commutative as a binary operation. Either `and` or `with` sound OK. Might be `append` as well.

Reply:
`append` is for appending single elements; `with` is weird to me as a Python programmer ;).

`merge`? `combine`?

Reply:
List append is an operation on two lists that returns a concatenation of the first list and second list. It's much older than Python, and Python got it differently (and wrong, to me as a functional programmer) with `extend` and `append`. Let's not use `append` to conform to the naming convention.

`merge` has a reordering aspect. `combine` or `combined` sound like the output doesn't preserve the exact shape of inputs. `join` is a synonym of "or" rather than `and`: https://en.wikipedia.org/wiki/Join_and_meet

I think `join` is the closest match out of the ones you listed.